### ISSUE TITLE
[Bug]: Performance regression caused by high-priority L2 Cache Data Persisting from Server Startup

### ISSUE THREAD

**Issue by xiaohongchen1991** [vllm-project/vllm#48518]
### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>

```text
Collecting environment information...
==============================
        System Info
==============================
OS                           : Ubuntu 22.04.5 LTS (x86_64)
GCC version                  : (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Clang version                : Could not collect
CMake version                : version 4.4.0
Libc version                 : glibc-2.35

==============================
       PyTorch Info
==============================
PyTorch version              : 2.11.0+cu130
Is debug build               : False
CUDA used to build PyTorch   : 13.0
ROCM used to build PyTorch   : N/A
XPU used to build PyTorch    : N/A

==============================
      Python Environment
==============================
Python version               : 3.12.13 (main, Jun 23 2026, 15:18:55) [Clang 22.1.3 ] (64-bit runtime)
Python platform              : Linux-5.15.0-185-generic-x86_64-with-glibc2.35

==============================
       CUDA / GPU Info
==============================
Is CUDA available            : True
CUDA runtime version         : Could not collect
CUDA_MODULE_LOADING set to   :
GPU models and configuration : GPU 0: NVIDIA H100 80GB HBM3
Nvidia driver version        : 590.48.01
cuDNN version                : Could not collect
HIP runtime version          : N/A
MIOpen runtime version       : N/A
Is XNNPACK available         : True
==============================
          CPU Info
==============================
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  20
On-line CPU(s) list:                     0-19
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) Platinum 8468
CPU family:                              6
Model:                                   143
Thread(s) per core:                      1
Core(s) per socket:                      20
Socket(s):                               1
Stepping:                                8
BogoMIPS:                                4200.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush\
 dts mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology cpu\
id tsc_known_freq pni pclmulqdq dtes64 vmx ssse3 fma cx16 pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer\
 aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single ssbd ibrs ibpb stibp ibrs_enh\
anced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq\
 rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx_vnni av\
x512_bf16 wbnoinvd arat avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg av\
x512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b fsrm md_clear serialize tsxldtrk avx512_fp16 arch_c\
apabilities
Virtualization:                          VT-x
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               640 KiB (20 instances)
L1i cache:                               640 KiB (20 instances)
L2 cache:                                80 MiB (20 instances)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-19
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Mitigation; Aligned branch/return thunks
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Unknown: No mitigations
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequenc\
e; BHI SW loop, KVM SW loop
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Mitigation; TSX disabled
Vulnerability Vmscape:                   Not affected


==============================
Versions of relevant libraries
==============================
[pip3] flashinfer-python==0.6.13
[pip3] helion==1.1.0
[pip3] numpy==2.3.5
[pip3] nvidia-cublas==13.1.0.3
[pip3] nvidia-cuda-cccl==13.3.3.4.1
[pip3] nvidia-cuda-crt==13.3.73
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvcc==13.2.78
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cuda-tileiras==13.2.78
[pip3] nvidia-cudnn-cu13==9.19.0.56
[pip3] nvidia-cudnn-frontend==1.26.0
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-cufile==1.15.1.6
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.0
[pip3] nvidia-cutlass-dsl==4.5.2
[pip3] nvidia-cutlass-dsl-libs-base==4.5.2
[pip3] nvidia-cutlass-dsl-libs-cu13==4.5.2
[pip3] nvidia-ml-py==13.610.43
[pip3] nvidia-nccl-cu13==2.28.9
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvshmem-cu13==3.4.5
[pip3] nvidia-nvtx==13.0.85
[pip3] nvidia-nvvm==13.2.78
[pip3] pyzmq==27.1.0
[pip3] tokenspeed-triton==3.8.10.post20260709
[pip3] torch==2.11.0
[pip3] torch_c_dlpack_ext==0.1.5
[pip3] torchaudio==2.11.0
[pip3] torchcodec==0.14.0
[pip3] torchvision==0.26.0
[pip3] transformers==5.13.1
[pip3] triton==3.6.0
[conda] Could not collect

==============================
         vLLM Info
==============================
ROCM Version                 : Could not collect
vLLM Version                 : 0.1.dev18637+g5edf5df55 (git sha: 5edf5df55)
vLLM Build Flags:
  CUDA Archs: Not Set; ROCm: Disabled; XPU: Disabled
GPU Topology:
        GPU0    CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      0-19    0               N/A

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

==============================
     Environment Variables
==============================
PYTORCH_NVML_BASED_CUDA_CHECK=1
TORCHINDUCTOR_COMPILE_THREADS=1
TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_shadeform
```

</details>

### 🐛 Describe the bug

Check out vLLM v0.25.1 and start up the server
```
python3 -m vllm.entrypoints.openai.api_server --model RedHatAI/Qwen3-8B-FP8-dynamic --max-num-seqs 4 --tensor-parallel-size 1 --no-enable-prefix-caching
```

In another terminal, run ```vllm bench serve```:
```
vllm bench serve \
  --backend vllm \
  --model RedHatAI/Qwen3-8B-FP8-dynamic \
  --endpoint /v1/completions \
  --dataset-name sharegpt \
  --dataset-path ShareGPT_V3_unfiltered_cleaned_split.json \
  --max-concurrency 4 \
  --num-warmups 32 \
  --ignore-eos \
  --num-prompts 512
```
The results returned:
```
============ Serving Benchmark Result ============
Successful requests:                     512
Failed requests:                         0
Maximum request concurrency:             4
Benchmark duration (s):                  138.96
Total input tokens:                      105324
Total generated tokens:                  112477
Request throughput (req/s):              3.68
Output token throughput (tok/s):         809.40
Peak output token throughput (tok/s):    844.00
Peak concurrent requests:                12.00
Total token throughput (tok/s):          1567.33
---------------Time to First Token----------------
Mean TTFT (ms):                          21.04
Median TTFT (ms):                        19.12
P99 TTFT (ms):                           33.70
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          4.83
Median TPOT (ms):                        4.81
P99 TPOT (ms):                           5.34
---------------Inter-token Latency----------------
Mean ITL (ms):                           4.84
Median ITL (ms):                         4.76
P99 ITL (ms):                            11.73
==================================================
```
Note that the mean TPOT (ms) is around 4.83. You can rerun this multiple times and ```vllm bench serve``` will return very close perf numbers.

Now, clean up the L2 cache using 
```
#!/usr/bin/env python3                                                                                                   

import torch
from triton import runtime


def main():
    if not torch.cuda.is_available():
        raise RuntimeError("CUDA is not available.")

    # Allocate the benchmark cache-clearing buffer once.                                                                 
    cache = runtime.driver.active.get_empty_cache_for_benchmark()

    print("Clearing GPU L2 cache...")
    runtime.driver.active.clear_cache(cache)
    torch.cuda.synchronize()
    print("Done.")


if __name__ == "__main__":
    main()
```

Rerunning ```vllm bench serve``` will return better perf numbers, 4.66 ms TPOT vs 4.83 ms before L2 cache cleanup.
```
============ Serving Benchmark Result ============
Successful requests:                     512
Failed requests:                         0
Maximum request concurrency:             4
Benchmark duration (s):                  133.94
Total input tokens:                      105324
Total generated tokens:                  112477
Request throughput (req/s):              3.82
Output token throughput (tok/s):         839.75
Peak output token throughput (tok/s):    872.00
Peak concurrent requests:                15.00
Total token throughput (tok/s):          1626.10
---------------Time to First Token----------------
Mean TTFT (ms):                          20.27
Median TTFT (ms):                        18.20
P99 TTFT (ms):                           33.05
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          4.66
Median TPOT (ms):                        4.64
P99 TPOT (ms):                           5.31
---------------Inter-token Latency----------------
Mean ITL (ms):                           4.66
Median ITL (ms):                         4.59
P99 ITL (ms):                            11.59
==================================================
```

Note that 
- If I clean up L2 cache immediately after vLLM server is ready, I will get the better perf numbers. So this confirms there are some data created during vLLM startup and the data is set as high-priority and hence persisting in L2 cache. 
- I tried running the same "vllm bench serve" session for like 5-10 times (each with 512 requests) and all get the worse perf numbers. So, it seems like the impact can persist for quite some time and the L2 cache state may not recover.

I didn't try to identify what data creation during server startup caused this. But I think it would be good to clear L2 cache right before server startup completion to address this issue.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by Sahil170595** [comment]
We attempted a controlled replication of this on an H100 80GB HBM3 (CUDA 13.0, vLLM `6f91edf96`, torch 2.13.0+cu130, triton 3.7.1) with the method from the report: `RedHatAI/Qwen3-8B-FP8-dynamic` (revision `05233ce`), ShareGPT_V3, 512 prompts per run, `--max-concurrency 4`, `--num-warmups 32`, `ignore_eos`.

Design differences from the report, both aimed at isolating the claimed effect:

1. **Randomized interleaving.** 10 blocks × 3 conditions with a fresh server per arm and balanced random order within each block, instead of A-runs followed by one B-run. (In the original design A was never re-run after B, so machine-state drift in either direction is indistinguishable from the intervention.)
2. **In-context interventions.** The persisting-L2 state of a CUDA context is not reachable from a separate cleanup process, so both interventions were executed *inside the serving worker's CUDA context* via `--worker-extension-cls` + `collective_rpc`: (a) `cudaCtxResetPersistingL2Cache()` — the API CUDA provides for exactly the state this issue describes — and (b) Triton's `clear_cache` (256 MiB write sweep), i.e. this issue's proposed remedy.

**Result: null.** Paired within-block deltas (positive = intervention better):

| Comparison | mean TPOT Δ% | median | positive blocks | mean tput Δ% | positive blocks |
|---|---:|---:|---:|---:|---:|
| reset vs untouched | −0.033 | −0.059 | 4/10 | −0.115 | 1/10 |
| evict (`clear_cache`) vs untouched | −0.026 | −0.038 | 4/10 | −0.067 | 3/10 |

The reported +3.5% TPOT / +3.75% throughput effect is two orders of magnitude larger than anything we observe; our deltas are sign-mixed and centred at zero.

Mechanism checks, independent of the benchmark:

- **No persisting-L2 host API is called during vLLM startup.** A CUPTI callback trace across 279 startup traces (cold profile through HTTP-200 readiness, captured on an RTX 4080 environment) recorded zero calls to `cudaDeviceSetLimit(cudaLimitPersistingL2CacheSize)`, `cuCtxSetLimit`, `cudaStreamSetAttribute`/`cuStreamSetAttribute` (access-policy windows), graph-node access-policy attributes, or the reset APIs, with zero dropped records. Static audit of vLLM, PyTorch, CUTLASS, FlashInfer, NCCL core, and DeepGEMM likewise finds no host-side persisting-L2 call.
- **The nonzero `cudaDeviceGetLimit` reading is a driver default, not startup state.** A live vLLM worker on the H100 reads `cudaLimitPersistingL2CacheSize = 9,830,400` — exactly 3/16 of its 50 MB L2, which at first looked like startup-established state. But a bare context created directly through the driver API with ctypes (no torch, no libraries loaded) on an RTX 4080 reads 9,437,184 — exactly 3/16 of *its* L2, identically on Windows and WSL2. Any fresh CUDA process on a CUDA-13-era driver reports this fraction; server startup did not establish it.
- **The remedy cannot act on the proposed mechanism.** Triton's `clear_cache` is a normal-priority write sweep ([`cache.zero_()`](https://github.com/triton-lang/triton/blob/main/third_party/nvidia/backend/driver.py)), and per the [CUDA programming guide](https://docs.nvidia.com/cuda/cuda-programming-guide/04-special-topics/l2-cache-control.html) normal accesses cannot evict persisting lines. Separately, a standalone cleanup process cannot touch the server context's persisting-L2 state at all.
- **Warmup arithmetic.** With `--num-warmups 32` and this model, ~58 TB of weight traffic flows through the 50 MB L2 before the first measured token (~1.2M full cache turnovers). Decode at batch 4 streams ~8.2 GB of weights per step (~164× the L2), so L2 contents are not a lever for steady-state TPOT on this workload regardless of startup state.

The most likely explanation for the original observation is the uncontrolled A-then-B design on a shared KVM guest: 5–10 consecutive ~140 s runs heat-soak the part, the intervention adds an idle gap (fresh process + torch import + CUDA init) that restores boost headroom, and TPOT is measured client-side where a 0.17 ms delta sits within asyncio jitter. Re-running the untouched arm *after* the intervention would discriminate — under randomized interleaving the effect disappears.

Happy to share the full harness (worker extension, randomization schedule, per-arm manifests, CUPTI tracer) if useful. Based on the above we'd suggest closing as not-reproducible unless a randomized, interleaved run on the original machine shows otherwise.
